### PR TITLE
Disallow MERGE of entities with null properties

### PR DESCRIPTION
**Pull request by jeffreylovitz**
Add appropriate restriction to MERGE creations and exit early on encountering invalid property errors.

### CODE REVIEW

File: src/execution_plan/ops/shared/create_functions.c [hunk 1]
```diff
@@ -165,7 +165,10 @@ PendingProperties *ConvertPropertyMap(Record r, const PropertyMap *map) {
 		converted->values[i] = AR_EXP_Evaluate(map->values[i], r);
 		// Emit an error and return NULL if we're trying to add an invalid type.
 		if(!(SI_TYPE(converted->values[i]) & SI_VALID_PROPERTY_VALUE)) {
+			converted->property_count = i;
+			PendingPropertiesFree(converted);
 			Error_InvalidPropertyValue();
+			QueryCtx_RaiseRuntimeException();
 			break;
```
Comment (swilly22) [inline src/execution_plan/ops/shared/create_functions.c:172]: Is this block covered by a test? I can see that MERGE is covered but I'm not sure about CREATE.

File: src/execution_plan/ops/op_update.c [hunk 2]
```diff
@@ -260,9 +260,10 @@ static void _EvalEntityUpdates(EntityUpdateCtx *ctx, GraphContext *gc,
 		EntityUpdateEvalCtx *update_ctx = ctx->exps + i;
 		SIValue new_value = AR_EXP_Evaluate(update_ctx->exp, r);
 
-		// Emit an error and return NULL if we're trying to add an invalid type.
+		// Emit an error and exit if we're trying to add an invalid type.
 		if(!(SI_TYPE(new_value) & SI_VALID_PROPERTY_VALUE)) {
 			Error_InvalidPropertyValue();
+			QueryCtx_RaiseRuntimeException();
```
Comment (swilly22) [inline src/execution_plan/ops/op_update.c:266]: Just as a suggestion, maybe rename and move `QueryCtx_RaiseRuntimeException` to `Error_RaiseRuntimeException`

File: src/execution_plan/ops/op_merge_create.c [hunk 3]
```diff
@@ -102,7 +102,16 @@ static bool _CreateEntities(OpMergeCreate *op, Record r) {
...
-		if(map) converted_properties = ConvertPropertyMap(r, map);
+		if(map) {
+			converted_properties = ConvertPropertyMap(r, map);
+			for(int i = 0; i < converted_properties->property_count; i++) {
+				if(SI_TYPE(converted_properties->values[i]) & T_NULL) {
+					PendingPropertiesFree(converted_properties);
+					QueryCtx_SetError("Cannot merge node using null property value");
+					QueryCtx_RaiseRuntimeException();
+				}
+			}
```
Comment (swilly22) [inline src/execution_plan/ops/op_merge_create.c:113]: Isn't this redundant ?
`ConvertPropertyMap` in `src/execution_plan/ops/shared/create_functions.c`  performs this validation already.

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: src/execution_plan/ops/op_merge_create.c (hunk 3 above)
Comment (jeffreylovitz) [reply src/execution_plan/ops/op_merge_create.c:113]: I don't know what the justification for these differing behaviors are, but:
```
CREATE (a:L {v1: NULL, v2: 'prop'}) RETURN a
```
Creates one node with the property `v2`, while:
```
MERGE (a:L {v1: NULL, v2: 'prop'}) RETURN a
```
Is supposed to generate this error.

**Review by swilly22 — CHANGES_REQUESTED** [review]
Let's continue online

File: src/graph/entities/graph_entity.c [hunk 4]
```diff
@@ -98,8 +100,8 @@ SIValue *GraphEntity_GetProperty(const GraphEntity *e, Attribute_ID attr_id) {
 void GraphEntity_SetProperty(const GraphEntity *e, Attribute_ID attr_id, SIValue value) {
 	ASSERT(e);
 
-	// Emit an error and return if we're trying to add an invalid type.
-	if(!_GraphEntity_ValidatePropertyValue(value)) return;
+	// Emit an error and exit if we're trying to add an invalid type.
+	_GraphEntity_ValidatePropertyValue(value);
```
[the remark this answers is not in the document]
Comment (jeffreylovitz) [reply src/graph/entities/graph_entity.c:104]: This call will still produce an error if the property value is a node or edge.

File: src/execution_plan/ops/shared/create_functions.c [hunk 5]
```diff
@@ -157,19 +157,45 @@ void CommitNewEntities(OpBase *op, PendingCreations *pending) {
...
+			 * otherwise skip this value. */
+			if(fail_on_null) {
+				// Emit an error and exit.
+				converted->property_count = i;
+				PendingPropertiesFree(converted);
+				ErrorCtx_RaiseRuntimeException("Cannot merge node using null property value");
+			} else {
+				// Replace the NULL value with the final conversion and reset the loop.
+				Attribute_ID *keys = (Attribute_ID *)map->keys;
+				AR_EXP_Free(map->values[i]);
```
Comment (swilly22) [inline src/execution_plan/ops/shared/create_functions.c:184]: `CREATE (), ( {v:1})`
`MATCH (n) CREATE ({x:n.v})`

Assuming the fist `n` node resolves to the empty node, as such `n.v` is NULL, which will remove the expressions `n.v` from the shared property map, causing the second created node to miss its `x` attribute.

**Review by swilly22 — CHANGES_REQUESTED** [review]

**Review by swilly22 — APPROVED** [review]

**Pull request merged**